From cd4df76cf17f873d57c21f4d0382a8c1a3339d5c Mon Sep 17 00:00:00 2001
From: Guy Nisan <gnisan@habana.ai>
Date: Mon, 08 Mar 2021 08:48:54 +0200
Subject: [PATCH] [SW-30456] boot info: set fit boot failure

In case that uboot failed to load the linux FIT, it will store the
reason for the failure in boot info.

This commit reads this info and sets CPU_BOOT_ERR0_x accordingly.

Change-Id: I715555358d0850e66050c6d1e3617aad974f8e46
Signed-off-by: Guy Nisan <gnisan@habana.ai>
---

diff --git a/include/boot_info.h b/include/boot_info.h
index 9d5558e..abf60cf 100644
--- a/include/boot_info.h
+++ b/include/boot_info.h
@@ -11,5 +11,6 @@
 
 void boot_info_finalize(void);
 int boot_info_set(u32 type, const void *data, u32 size);
+int boot_info_get(u32 type, void *data, u32 size);
 
 #endif /* _BOOT_INFO_H_ */
diff --git a/preboot/boot_info.c b/preboot/boot_info.c
index 4e27182..a5421ec 100644
--- a/preboot/boot_info.c
+++ b/preboot/boot_info.c
@@ -139,3 +139,49 @@
 
 	return 0;
 }
+
+int boot_info_get(u32 type, void *data, u32 size)
+{
+	struct hl_boot_info_ptrs ptrs;
+	struct hl_boot_info_desc desc;
+	u32 *offsets, *src, *dst, i, j;
+	u8 chksum[HL_CHKSUM_SIZE];
+
+	/* Size must be 32bit aligned */
+	if (size & (sizeof(u32) - 1))
+		return 1;
+
+	dst = data;
+
+	ptrs = *(struct hl_boot_info_ptrs *)offset2ptr(0, BOOT_INFO_BASE);
+	if (ptrs.magic != HL_BOOT_INFO_MAGIC)
+		return 1;
+
+	offsets = (u32 *)offset2ptr(sizeof(struct hl_boot_info_ptrs),
+				    BOOT_INFO_BASE);
+
+	for (i = 0; i < ptrs.count; i++) {
+		desc = *(struct hl_boot_info_desc *)offset2ptr(offsets[i],
+							       BOOT_INFO_BASE);
+		if (type != desc.type)
+			continue;
+
+		if (desc.magic != (HL_BOOT_INFO_DESC_MAGIC | type))
+			return 1;
+
+		if (size < desc.size)
+			return 1;
+
+		src = offset2ptr(desc.offset, BOOT_INFO_BASE);
+		for (j = 0; j < ROUND_UP4(desc.size) / sizeof(u32); j++)
+			dst[j] = src[j];
+
+		sha256_csum((u8 *)data, desc.size, chksum);
+		if (memcmp(chksum, desc.chksum, HL_CHKSUM_SIZE))
+			return 1;
+
+		return 0;
+	}
+
+	return 1;
+}
diff --git a/preboot/preboot.c b/preboot/preboot.c
index 85e1e1c..aca9ab2 100644
--- a/preboot/preboot.c
+++ b/preboot/preboot.c
@@ -254,6 +254,31 @@
 	device_id_vendor_id = regmap_readl(mmPCIE_DBI_DEVICE_ID_VENDOR_ID_REG);
 }
 
+static void preboot_process_last_boot_info(void)
+{
+	u32 boot_err = HL_FW_FIT_BOOT_ERR_NONE;
+
+	if (!boot_info_get(HL_BOOT_INFO_FIT_BOOT_ERROR, &boot_err,
+			   sizeof(boot_err))) {
+		switch (boot_err) {
+		case HL_FW_FIT_BOOT_ERR_NONE:
+			break;
+		case HL_FW_FIT_BOOT_ERR_CORRUPTED:
+		case HL_FW_FIT_BOOT_ERR_ADDRESS_VIOLATION:
+		case HL_FW_FIT_BOOT_ERR_NOT_RUNNABLE:
+			cpu_boot_err_clrset(CPU_BOOT_ERR0_FIT_CORRUPTED, true);
+			break;
+		case HL_FW_FIT_BOOT_ERR_REVOKED_KEY:
+		case HL_FW_FIT_BOOT_ERR_SVN:
+		case HL_FW_FIT_BOOT_ERR_SIGNATURE:
+			cpu_boot_err_clrset(CPU_BOOT_ERR0_SECURITY_FAIL, true);
+			break;
+		}
+
+		return;
+	}
+}
+
 __weak void hl_dev_pll_init(void) {}
 __weak void hl_clock_gating_init(void) {}
 __weak void preboot_conf_iatu_access(bool enable) {}
@@ -305,5 +330,6 @@
 	hl_ram_init();
 	preboot_dev_enable_caps();
 	crc32_init_lookup();
+	preboot_process_last_boot_info();
 	module_handle_cmds(cmd_ops, ARRAY_SIZE(cmd_ops));
 }
